jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java

```diff
@@ -50,7 +50,9 @@ public class AsyncContent implements Content.Sink, Content.Source, Closeable
     @Override
     public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
     {
-        write(Content.Chunk.from(byteBuffer, last, callback::succeeded), callback);
+        boolean terminal = !byteBuffer.hasRemaining() && last;
+        Content.Chunk chunk = terminal ? Content.Chunk.EOF : Content.Chunk.from(byteBuffer, last, callback::succeeded);
```

Comment:
Doesn't this imply that callback succeeded will get called twice? Once for the chunk release and again when the write succeeds?

Reply:
Hmmm it always did... Still a good question?

Reply:
You're right: it was and still is a bug. Let me fix this.